iocore/net/P_SNIActionPerformer.h

```diff
@@ -92,180 +95,62 @@ class HTTP2BufferWaterMark : public ActionItem
-    } else if (has_recv_port_var) {
-      fnArrIndexes.push_back(OpId::MAP_WITH_RECV_PORT);
-      var_start_pos = recv_port_start_pos;
-    } else if (has_pp_port_var) {
-      fnArrIndexes.push_back(OpId::MAP_WITH_PROXY_PROTOCOL_PORT);
-      var_start_pos = pp_port_start_pos;
-    }
-    // Check for match groups as well.
-    if (destination.find_first_of('$') != std::string::npos) {
-      fnArrIndexes.push_back(OpId::MATCH_GROUPS);
-    }
-  }
-  ~TunnelDestination() override {}
+  public:
+    CaptureArgs(std::optional<ActionItem::Context::CapturedGroupViewVec> const &groups);
 
-  int
-  SNIAction(TLSSNISupport *snis, const Context &ctx) const override
-  {
-    // Set the netvc option?
-    SSLNetVConnection *ssl_netvc = dynamic_cast<SSLNetVConnection *>(snis);
-    const char *servername       = snis->get_sni_server_name();
-    if (ssl_netvc) {
-      if (fnArrIndexes.empty()) {
-        ssl_netvc->set_tunnel_destination(destination, type, !TLSTunnelSupport::PORT_IS_DYNAMIC, tunnel_prewarm);
-        Debug("ssl_sni", "Destination now is [%s], fqdn [%s]", destination.c_str(), servername);
-      } else {
-        bool port_is_dynamic = false;
-        auto fixed_dst{destination};
-        // Apply mapping functions to get the final destination.
-        for (auto fnArrIndex : fnArrIndexes) {
-          // Dispatch to the correct tunnel destination port function.
-          fixed_dst = fix_destination[fnArrIndex](fixed_dst, var_start_pos, ctx, ssl_netvc, port_is_dynamic);
-        }
-        ssl_netvc->set_tunnel_destination(fixed_dst, type, port_is_dynamic, tunnel_prewarm);
-        Debug("ssl_sni", "Destination now is [%s], configured [%s], fqdn [%s]", fixed_dst.c_str(), destination.c_str(), servername);
-      }
+    virtual std::any capture(unsigned idx) const override;
 
-      if (type == SNIRoutingType::BLIND) {
-        ssl_netvc->attributes = HttpProxyPort::TRANSPORT_BLIND_TUNNEL;
-      }
+    /// Call out from formatting when a replace group is referenced.
+    virtual swoc::BufferWriter &print(swoc::BufferWriter &w, swoc::bwf::Spec const &spec, unsigned idx) const override;
 
-      // ALPN
-      for (int id : alpn_ids) {
-        ssl_netvc->enableProtocol(id);
-      }
-    }
+    /// Number of arguments in the pack.
+    virtual unsigned count() const override;
 
-    return SSL_TLSEXT_ERR_OK;
-  }
+  private:
+    /// Empty group for when no captures exist.
+    static inline const ActionItem::Context::CapturedGroupViewVec NO_GROUPS;
+    ActionItem::Context::CapturedGroupViewVec const &_groups;
+  };
 
-private:
-  static bool
-  is_number(std::string_view s)
-  {
-    return !s.empty() &&
-           std::find_if(std::begin(s), std::end(s), [](std::string_view::value_type c) { return !std::isdigit(c); }) == std::end(s);
-  }
+  static constexpr std::string_view MAP_WITH_RECV_PORT_STR           = "inbound_local_port";
+  static constexpr std::string_view MAP_WITH_PROXY_PROTOCOL_PORT_STR = "proxy_protocol_port";
```

Comment:
In the current code, specifying both the `inbound_local_port` and `proxy_protocol_port` results in an error. Looking at the changes it seems all available substitutions are applied by default. Is it possible to preserve the error for this misconfigured case?

Reply:
Is that misconfigured? Currently it's a misconfiguration because one of them will not be applied and therefore the behavior is not as expected. With this change all instances are replaced and therefore there is no surprise if both are used. I don't see any reason to forbid it.

Reply:
@brbzull0 Do you think we should support this? If so, the doc would need to be updated.